File: src/spark/query_2.py
from typing import List, Tuple
from pyspark.sql import DataFrame, GroupedData, SparkSession
from pyspark.sql.functions import col, collect_set, concat_ws, count, window


def __query(df: GroupedData) -> DataFrame:
    return (
        # aggregate by failure, models and serial_numbers
        df.agg(
            count("failure").alias("failures"),
            collect_set(
                concat_ws(
                    ",",
                    col("model"),
                    col("serial_number"),
                )
            ).alias("failed_hdds"),
        )
        # order by failures
        .orderBy(col("failures").desc())
        # limit to ranking
        .limit(10)
    )


def __format(df: DataFrame, prefix: str, spark: SparkSession):
    # get all the item from the batch
    rows = df.collect()
    # create output rows with the initial date of the batch
    output_rows = [rows[0]["date"]]
    # iterate through all the rows (ranking positions)
    for row in rows:
        # create list of failed hard disks
        hdds = ",".join(row["failed_hdds"])
        # output string of vault (with failures and failed hdds)
        ranking = f"{row["vault_id"]},{row["failures"]}({hdds})"
        # add ranking to all rows
        output_rows.append(ranking)

    # create dataframe containing a single item by joining the ranking in a single line
    output_df = spark.createDataFrame([(",".join(output_rows),)])
    # write as text to output file
    output_df.write.mode("append").text(f"/app/output/{prefix}")


def save_query_2(wnds: List[Tuple[DataFrame, str]], spark: SparkSession):
    for wnd, prefix in wnds:
        (
            # process stream
            wnd.writeStream
            # in complete mode
            .outputMode("complete")
            # format and write output batch
            .foreachBatch(lambda df, _: __format(df, prefix, spark))
            .start()
        )


def query_2(df: DataFrame) -> List[Tuple[DataFrame, str]]:
    # filter only failures
    filtered_df = df.filter(df["failure"] > 0)

    window_1_day = (
        # set window size (1 day)
        filtered_df.withWatermark("date", "1 day")
        # group by window size and vault_id
        .groupBy(window("date", "1 day"), "vault_id")
    )
    window_3_days = (
        # set window size (1 day)
        filtered_df.withWatermark("date", "3 days")
        # group by window size and vault_id
        .groupBy(window("date", "3 days"), "vault_id")
    )

    # group only by vault_id
    window_all = filtered_df.groupBy("vault_id")

    return [
        (__query(window_1_day), "query_1_day_1"),
        (__query(window_3_days), "query_1_days_3"),
        (__query(window_all), "query_1_days_all"),
    ]

File: src/spark/src/main/scala/queries/Query2.scala
package queries

import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.Row
import org.apache.spark.sql.functions._

object Query2 {
  private def impl(df: Dataset[Row], wnd: Long, offset: Long): DataFrame = {
    return df
      .groupBy(
        window(col("date_ts"), s"$wnd days", s"$wnd days", s"$offset days"),
        col("vault_id")
      )
      // calculate sum of failures and list of (model, serial_number)
      .agg(
        count("failure").alias("failures"),
        collect_set(concat_ws(",", col("model"), col("serial_number")))
          .as("hdds")
      )
      // order by failures
      .orderBy(desc("failures"))
      // create ranking
      .limit(10)
      // select ts, vault_id, failures, hdds
      .select(
        date_format(col("window.start"), "yyyy-MM-dd'T'HH:mm:ss.SSSSSS")
          .as("ts"),
        col("vault_id"),
        col("failures"),
        col("hdds")
      )
      // group by timestamp
      .groupBy("ts")
      // aggregate all the vaults with the same timestamp into a list
      .agg(
        collect_list(struct(col("vault_id"), col("failures"), col("hdds")))
          .as("rankings")
      )
      // select as ts,ranking
      .select(
        col("ts") +:
          (1 to 10).flatMap(i =>
            Seq(
              expr(s"if(size(rankings) >= $i, rankings[${i - 1}].vault_id, '')")
                .alias(s"vault_id_$i"),
              expr(s"if(size(rankings) >= $i, rankings[${i - 1}].failures, 0)")
                .alias(s"failures_$i"),
              expr(
                s"if(size(rankings) >= $i, rankings[${i - 1}].hdds, array())"
              ).alias(s"hdss$i")
            )
          ): _*
      )
  }

  def query(df: Dataset[Row]): List[(Dataset[Row], String)] = {
    // Filter only vaults with failures
    val filtered_df = df
      .filter(col("failure") > 0)
      // select only the necessary columns
      .select("date_ts", "vault_id", "failure", "model", "serial_number")

    return List(
      (impl(filtered_df, 1, 0), "query_1"),
      (impl(filtered_df, 3, 2), "query_3"),
      (impl(filtered_df, 23, 13), "query_23")
    )
  }
}
